x/accounts/vesting/periodic_vesting_account.go

```
@@ -0,0 +1,257 @@
+func (pva PeriodicVestingAccount) IteratePeriods(
+	ctx context.Context,
+	cb func(_ string, value vestingtypes.Period) bool,
+) {
+	err := pva.VestingPeriods.Walk(ctx, nil, func(key string, value vestingtypes.Period) (stop bool, err error) {
+		return cb(key, value), nil
+	})
+	if err != nil {
+		panic(err)
+	}
+}
+
+// GetVestedCoins returns the total number of vested coins. If no coins are vested,
+// nil is returned.
+func (pva PeriodicVestingAccount) GetVestCoinsInfo(ctx context.Context, blockTime time.Time) (vestedCoins, vestingCoins sdk.Coins, err error) {
+	vestedCoins = sdk.Coins{}
+	vestingCoins = sdk.Coins{}
+
+	// We must handle the case where the start time for a vesting account has
+	// been set into the future or when the start of the chain is not exactly
+	// known.
+	startTime, err := pva.StartTime.Get(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	endTime, err := pva.EndTime.Get(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	originalVesting := sdk.Coins{}
+	pva.IterateCoinEntries(ctx, pva.OriginalVesting, func(key string, value math.Int) (stop bool) {
+		originalVesting = append(originalVesting, sdk.NewCoin(key, value))
+		return false
+	})
+	if math.NewInt(blockTime.Unix()).LTE(startTime) {
+		return vestedCoins, originalVesting, nil
+	} else if math.NewInt(blockTime.Unix()).GTE(endTime) {
+		return originalVesting, vestingCoins, nil
+	}
+
+	// track the start time of the next period
+	currentPeriodStartTime, err := pva.StartTime.Get(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	pva.IteratePeriods(ctx, func(_ string, period vestingtypes.Period) (stop bool) {
+		x := math.NewInt(blockTime.Unix()).Sub(currentPeriodStartTime).Int64()
+		if x < period.Length {
+			return true
+		}
+
+		vestedCoins = vestedCoins.Add(period.Amount...)
+
+		// update the start time of the next period
+		err = pva.StartTime.Set(ctx, currentPeriodStartTime.Add(math.NewInt(period.Length)))
+		return err != nil
```

the error is ignored, we should return an error